### PR TITLE
setup cmake

### PR DESCRIPTION
**Pull request by linkerzhang**
This brings convenience for devs like me (coding with vs) in windows. :-)

c++ test (based on google test) will be added later if we all agree.

### CODE REVIEW

**Comment by bddppq** [comment]
Good. I think it makes senses to setup a cmake build system for building the c++ part in onnx, as we lean to have more stuffs in c++ to support more execution environments, and soon we want to provide  an onnx c++ library (without the python part).
On the other hand, having two ways (setup.py and cmake) to build onnx will eventually create divergence and it's hard to maintain. I would suggest let's use cmake as the source of truth to build onnx c++ part and change setup.py to use it.

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Minimum CMake required
+cmake_minimum_required(VERSION 3.5)
```
Comment (Yangqing) [inline cmake/CMakeLists.txt]: Might be a bit strict for build platforms like ubuntu 14.04.

File: cmake/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+# Project
+project(onnxir C CXX)
+
+# Set C++14 as standard for the whole project
+set(CMAKE_CXX_STANDARD 14)
```
Comment (Yangqing) [inline cmake/CMakeLists.txt]: Do we... really need this high?

IIRC Android support for c++14 is still semi-complete so I think a safe choice is 11, unless we are already stuck with 14.

File: cmake/external/protobuf.cmake [hunk 3]
```diff
@@ -0,0 +1,45 @@
+include (ExternalProject)
+
+set(PROTOBUF_INCLUDE_DIRS ${CMAKE_CURRENT_BINARY_DIR}/protobuf/src/protobuf/src)
```
Comment (Yangqing) [inline cmake/external/protobuf.cmake]: This pretty much hard-codes protobuf to be the in-house one; do we want to do something like FindProtobuf.cmake so that when someone is building protobuf, they can use the system installed one?

File: cmake/external/zlib.cmake [hunk 4]
```diff
@@ -0,0 +1,47 @@
+include (ExternalProject)
+
+set(zlib_INCLUDE_DIR ${CMAKE_CURRENT_BINARY_DIR}/external/zlib_archive)
```
Comment (Yangqing) [inline cmake/external/zlib.cmake]: Do we need zlib? If I understand it correctly, this is only used as a downstream dependency for protobuf right?

File: cmake/patches/protobuf.patch [hunk 5]
```diff
@@ -0,0 +1,27 @@
+diff --git a/src/google/protobuf/compiler/cpp/cpp_file.cc b/src/google/protobuf/compiler/cpp/cpp_file.cc
```
Comment (Yangqing) [inline cmake/patches/protobuf.patch]: Um... I would refrain from patching a well known library; we should probably upstream this to google/protobuf.

File: cmake/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /WX")
+    set(CMAKE_STATIC_LINKER_FLAGS "${CMAKE_STATIC_LINKER_FLAGS} /WX")
+endif()
+
+include_directories(
+    ${ONNXIR_ROOT}
+    ${CMAKE_CURRENT_BINARY_DIR}
+    # External dependencies.
+    ${PROTOBUF_INCLUDE_DIRS}
+)
```
Comment (houseroad) [inline cmake/CMakeLists.txt]: Looks like no newline at the end.

File: cmake/patches/protobuf.patch (hunk 5 above)
Comment (Yangqing) [reply cmake/patches/protobuf.patch]: Actually, I would recommend adding /wd4800 to the compiler flag, instead of explicitly patching. Also, you will encounter way more than C4800 alone, see

https://github.com/google/protobuf/tree/master/cmake

(see "Notes on Compiler Warnings")

File: cmake/CMakeLists.txt (hunk 2 above)
Comment (houseroad) [reply cmake/CMakeLists.txt]: agree, 14 is probably too aggressive. Do we use any C++14 feature?

Comment (linkerzhang) [reply cmake/CMakeLists.txt]: thank you! 11 is good.

File: cmake/external/protobuf.cmake (hunk 3 above)
Comment (linkerzhang) [reply cmake/external/protobuf.cmake]: I personally think it's good to use a hard code version and pull it from source. Let me know if you have strong preferences here.

File: cmake/external/zlib.cmake (hunk 4 above)
Comment (linkerzhang) [reply cmake/external/zlib.cmake]: thank you! Yangqing. We don't need it. Removed.

File: cmake/patches/protobuf.patch (hunk 5 above)
Comment (linkerzhang) [reply cmake/patches/protobuf.patch]: OK. changed to disable the warning in the project level.

**Comment by linkerzhang** [comment]
@bddppq @Yangqing ,do you have more comments there please? for referring fixed version of protobuf, let me know if you have strong preferences please. For making setup.py using cmake, I agree on it and will make it (need to learn more about distutils :) ) in separate PR to move this forward so that some folks could get convenience by using the cmake.

**Comment by bddppq** [comment]
Yeah for protobuf we need to use the shared one, otherwise it can easily get version/symbols conflict with other libs that link with the system protobuf. 
Changing setup.py in a separate diff makes senses to me. (You can look at caffe2’s setup.py to see how to delegate the c++ part to cmake).

**Comment by linkerzhang** [comment]
@bddppq Thank you for the clarification! hmmm, are we saying we want to find and depend on the protobuf installed locally on each dev environment please? that means that people may use different version of protobuf. I think that's sth we don't want to lead to. No?

**Comment by yinghai** [comment]
Any blockers? I really would like see this to be pulled so that we can make integration of ONNX easier.

**Comment by bddppq** [comment]
@linkerzhang We need to use the protobuf in user's environment. When users use onnx, it's very likely they will use it with frameworks/libs that depends on protobuf. Importing two libs linked with different copies of protobuf can easily get you symbols conflicts and version mismatch.

**Comment by linkerzhang** [comment]
@bddppq OK, I'm now using local protobuf to push this in to unblock folks. For onnx-ml handling, I'll make it in a separate PR, btw.

File: onnx/onnx_pb.h [hunk 7]
```diff
@@ -7,7 +7,7 @@
 #ifdef ONNX_ML
 #include "onnx/onnx-ml.pb.h"
```
Comment (linkerzhang) [inline onnx/onnx_pb.h]: onnx-ml will be handled in a separate PR, depending on my abstract type PR.

File: cmake/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+set(ONNXIR_ROOT ${PROJECT_SOURCE_DIR}/..)
+if (WIN32)
+    # parallel build
+    add_compile_options(/MP)
+    add_compile_options(/WX)
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /WX")
+    set(CMAKE_STATIC_LINKER_FLAGS "${CMAKE_STATIC_LINKER_FLAGS} /WX")
+endif()
+add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/../third_party/pybind11 ${CMAKE_CURRENT_BINARY_DIR}/pybind11)
+INCLUDE(FindProtobuf)
```
Comment (Yangqing) [inline cmake/CMakeLists.txt]: Not sure if we need to include it - the following line should basically find protobuf already.

File: cmake/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+set(CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/external)
+set(ONNXIR_ROOT ${PROJECT_SOURCE_DIR}/..)
+if (WIN32)
+    # parallel build
+    add_compile_options(/MP)
+    add_compile_options(/WX)
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /WX")
+    set(CMAKE_STATIC_LINKER_FLAGS "${CMAKE_STATIC_LINKER_FLAGS} /WX")
+endif()
+add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/../third_party/pybind11 ${CMAKE_CURRENT_BINARY_DIR}/pybind11)
```
Comment (Yangqing) [inline cmake/CMakeLists.txt]: Since pybind11 is a header only library, we probably don't need a binary dir for add_subdirectory I assume?

File: onnx/onnx_pb.h [hunk 10]
```diff
@@ -7,7 +7,7 @@
...
 #else
-#include "onnx/onnx.pb.h"
+#include "onnx.pb.h"
```
Comment (bddppq) [inline onnx/onnx_pb.h]: Why changing this?

File: cmake/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+INCLUDE(FindProtobuf)
+FIND_PACKAGE(Protobuf REQUIRED)
+include_directories(
+    ${ONNXIR_ROOT}
+    ${CMAKE_CURRENT_BINARY_DIR}
+    # External dependencies.
+    ${PROTOBUF_INCLUDE_DIRS}
+    ${PYBIND11_INCLUDE_DIR}
+    ${PYTHON_INCLUDE_DIRS}
```
Comment (Yangqing) [inline cmake/CMakeLists.txt]: Just checking - since there are no find python command, PYTHON_INCLUDE_DIRS is automatically provided by pybind11, I assume?

File: cmake/onnx.cmake [hunk 12]
```diff
@@ -0,0 +1,34 @@
+PROTOBUF_GENERATE_CPP(PROTO_SRCS PROTO_HDRS
```
Comment (Yangqing) [inline cmake/onnx.cmake]: I am not sure if this is the convention, but usually files under cmake/ are utility files that cmake uses for custom builds. For the content of this file, it seems to be a better fit as a onnx/CMakeLists.txt file. Does that make sense?

File: cmake/onnx.cmake [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+file(GLOB_RECURSE onnx_exclude_src
+    "${ONNXIR_ROOT}/onnx/onnx.pb.h"
+    "${ONNXIR_ROOT}/onnx/onnx.pb.cc"
+    "${ONNXIR_ROOT}/onnx/onnx-operators.pb.h"
+    "${ONNXIR_ROOT}/onnx/onnx-operators.pb.cc"
+)
+list(LENGTH onnx_exclude_src list_len)
+if (list_len GREATER 1)
+    list(REMOVE_ITEM onnx_src ${onnx_exclude_src})
```
Comment (Yangqing) [inline cmake/onnx.cmake]: Do you mind adding a piece of comment about the need for the list removal?

File: cmake/onnx.cmake [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+endif()
+
+add_library(onnxir ${PROTO_SRCS} ${PROTO_HDRS} ${onnx_src})
+
+if (WIN32)
+    target_compile_options(onnxir PRIVATE
```
Comment (Yangqing) [inline cmake/onnx.cmake]: A friendly reminder that you might need more than these warnings, thanks to protobuf which is a huge pot of warning-generator. More specifically, this list might help:

https://github.com/caffe2/caffe2/blob/3eb636c0edc9337fb348ec554e366b1020604c68/cmake/MiscCheck.cmake#L110-L126

File: onnx/onnx_pb.h (hunk 10 above)
Comment (Yangqing) [inline onnx/onnx_pb.h]: Quick check if this is intentional. Seems that we might want to indeed to onnx/onnx.pb.h?

(If you made this change due to protobuf generated files, please ping back - I can help with that)

Comment (linkerzhang) [reply onnx/onnx_pb.h]: Aha, you're right, I made this change due to files generated by protoc were put directly in build folder. Didn't figure out a way here to tell PROTOBUF_GENERATE_CPP to put generated files in a specific folder :(

File: cmake/onnx.cmake (hunk 14 above)
Comment (linkerzhang) [reply cmake/onnx.cmake]: currently only the warnings that specified here need to be taken care of, as I can get a successful build now. We may add more when needed?

File: cmake/onnx.cmake (hunk 13 above)
Comment (linkerzhang) [reply cmake/onnx.cmake]: this is temporary thing, since setup.py is not changed to use cmake for now. The proto header/code files will also be generated by setup.py (pip install -e ./), so, I'm adding the temp logic here to remove them if they're generated. It should be removed later.

File: onnx/onnx_pb.h (hunk 10 above)
Comment (linkerzhang) [reply onnx/onnx_pb.h]: Aha, it's because that the header generated by protoc is put in /build folder directly. Yanqing asked the same question.

File: cmake/CMakeLists.txt (hunk 9 above)
Comment (linkerzhang) [reply cmake/CMakeLists.txt]: you're right. it's not needed to add a binary dir.

Because I'm now putting cmakelists.txt in cmake folder instead of root folder, it creates out of source tree case. In this case, cmake/add_subdirectory asks to put a binary dir. I personally don't want to move the cmakelists.txt into root folder btw.

File: cmake/onnx.cmake (hunk 12 above)
Comment (linkerzhang) [reply cmake/onnx.cmake]: Hmmm, I think it in this way, 1) put all cmake related files into /cmake folder, instead of having them in different folders. 2) we may have more than one *.cmake later for different project separations, Sounds good?

File: cmake/CMakeLists.txt (hunk 8 above)
Comment (linkerzhang) [reply cmake/CMakeLists.txt]: you're right, it's not needed, removed. Thank you!

File: cmake/CMakeLists.txt (hunk 11 above)
Comment (linkerzhang) [reply cmake/CMakeLists.txt]: yes.

File: cmake/onnx.cmake (hunk 12 above)
Comment (linkerzhang) [reply cmake/onnx.cmake]: OK, now I merged things in one CmakeLists.txt and put it in root folder, which resolved several of the comments here.

File: cmake/CMakeLists.txt (hunk 9 above)
Comment (linkerzhang) [reply cmake/CMakeLists.txt]: Resolved this now by moving the CmakeLists.txt file in root folder now.

File: onnx/onnx_pb.h (hunk 10 above)
Comment (Yangqing) [reply onnx/onnx_pb.h]: Yeah, so protobuf's default command does not allow that, which is a big bummer. You can use a custom command with something like this:

https://github.com/caffe2/caffe2/blob/4f534fad1af9f77d4f0496ecd37dafb382330223/cmake/ProtoBuf.cmake#L137-L184

Note that if you want to use onnx as Windows dlls, you are going to need the dllexport_decl definition because otherwise protobuf does not work well with dlls.

File: CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+)
+
+# Temp solution to remove files generated by calling 'pip install ..."
+# Should be removed after changing setup.py to use cmake to build c++ part.
+list(LENGTH onnx_exclude_src list_len)
+if (list_len GREATER 1)
+    list(REMOVE_ITEM onnx_src ${onnx_exclude_src})
+endif()
+
+add_library(onnxir ${PROTO_SRCS} ${PROTO_HDRS} ${onnx_src})
```
Comment (Yangqing) [inline CMakeLists.txt]: In theory, onnxir should also depend on the protobuf libraries (if we build it as libonnxir.so). This might only work for cases where BUILD_SHARED_LIBS=OFF.

File: CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,65 @@
+# Minimum CMake required
+cmake_minimum_required(VERSION 3.1)
+
+# Project
+project(onnxir C CXX)
+
+# Set C++11 as standard for the whole project
+set(CMAKE_CXX_STANDARD 11)
```
Comment (ezyang) [inline CMakeLists.txt]: Good, your cmake minimum required is 3.1 :)

File: onnx/onnx_pb.h (hunk 10 above)
Comment (linkerzhang) [reply onnx/onnx_pb.h]: Thank you Yangqing! Yep, that's my origin change to use a customized generation function. OK, changed to the way.

File: CMakeLists.txt (hunk 15 above)
Comment (linkerzhang) [reply CMakeLists.txt]: that should be added as a link dependency in application-layer cmake file, right?

File: CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+include_directories(
+    ${ONNXIR_ROOT}
+    ${CMAKE_CURRENT_BINARY_DIR}
+    # External dependencies.
+    ${PROTOBUF_INCLUDE_DIRS}
+    ${PYBIND11_INCLUDE_DIR}
+    ${PYTHON_INCLUDE_DIRS}
+)
+
+function(RELATIVE_PROTOBUF_GENERATE_CPP SRCS HDRS ROOT_DIR)
```
Comment (ezyang) [inline CMakeLists.txt]: We cribbed this function definition from somewhere, right? Let's record its provenance please.

File: CMakeLists.txt [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+RELATIVE_PROTOBUF_GENERATE_CPP(PROTO_SRCS PROTO_HDRS ${ONNXIR_ROOT}
+    onnx/onnx.proto
+)
+
+file(GLOB_RECURSE onnx_src
+    "${ONNXIR_ROOT}/onnx/*.h"
+    "${ONNXIR_ROOT}/onnx/*.cc"
+)
+
+file(GLOB_RECURSE onnx_exclude_src
```
Comment (ezyang) [inline CMakeLists.txt]: You don't actually need a Glob here, right?

File: CMakeLists.txt (hunk 15 above)
Comment (ezyang) [reply CMakeLists.txt]: It's better if you declare it in the cmake. Then if someone else recursively includes this cmake, and they depend on onnxir, they will automatically see the protobuf dependency. If you omit it here they won't.

File: CMakeLists.txt (hunk 17 above)
Comment (linkerzhang) [reply CMakeLists.txt]: Aha, that's sth from ms code base, let me ask people whether they know where they copied from.

File: CMakeLists.txt (hunk 18 above)
Comment (linkerzhang) [reply CMakeLists.txt]: I think I don't need recursive here. Anyway, I'm changing to use "Remove" directly. This piece of codes will be removed anyway later.

File: CMakeLists.txt (hunk 17 above)
Comment (linkerzhang) [reply CMakeLists.txt]: looks like I can't figure out the code owner for now. I searched around and https://github.com/tensorflow/tensorflow/blob/d2c3b873c6f8ff999a2e4ee707a84ff00d9c15a5/tensorflow/contrib/cmake/tf_core_framework.cmake has the most similar one, so I'm going to put this link as a comment in place, sounds good?

File: CMakeLists.txt (hunk 18 above)
Comment (bddppq) [reply CMakeLists.txt]: After changing setup.py you will still need this. We will only use setup.py to manage the python packaging part, while cmake is going to be the one that does all the heavy lifting (including invoking protoc).

**Comment by linkerzhang** [comment]
@bddppq what's onnx-fb-universe used for please? :)

**Comment by ezyang** [comment]
It's our integration testing for onnx/onnx-caffe2/onnx-pytorch :)

**Comment by linkerzhang** [comment]
Aha, is there a magic command to ask the bot (onnx-fb-universe) to re-trigger the build please?

**Comment by houseroad** [comment]
@onnxbot retest this please?

**Comment by bddppq** [comment]
@houseroad the magic command needs to be done in the onnx-fb-universe repo :-)

@dzhulgakov @lupesko @prasanthpul Could you guys adjust the configuration to mark onnx-fb-universe CI as non-blocking for being able to merge a PR, or entirely remove the "Require status checks to pass before merging" mark. This is the second time CI outage being a blocker for merging PRs since we have enabled this check last week. I don't think any existing CI is robust enough (well, probably except the "license/cla" one).

File: CMakeLists.txt [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+    message(${FIL})
+    set(ABS_FIL ${ROOT_DIR}/${FIL})
+    get_filename_component(FIL_WE ${FIL} NAME_WE)
+    get_filename_component(FIL_DIR ${ABS_FIL} DIRECTORY)
+    file(RELATIVE_PATH REL_DIR ${ROOT_DIR} ${FIL_DIR})
+
+    list(APPEND ${SRCS} "${CMAKE_CURRENT_BINARY_DIR}/${REL_DIR}/${FIL_WE}.pb.cc")
+    list(APPEND ${HDRS} "${CMAKE_CURRENT_BINARY_DIR}/${REL_DIR}/${FIL_WE}.pb.h")
+
+    add_custom_command(
```
Comment (Yangqing) [inline CMakeLists.txt]: Note that if you want to build dlls, you will want to handle that with dllexport symbols. This is how we do it in c2:

https://github.com/caffe2/caffe2/blob/b7d983f255ef5496474f1ea188edb5e0ac442761/cmake/ProtoBuf.cmake#L165-L184

Not sure if I pointed that out before, so I'll just put it up here again just in case.

File: CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+endfunction()
+
+RELATIVE_PROTOBUF_GENERATE_CPP(PROTO_SRCS PROTO_HDRS ${ONNXIR_ROOT}
+    onnx/onnx.proto
+)
+
+# Temp solution to remove files generated by calling 'pip install ..."
+# Should be removed after changing setup.py to use cmake to build c++ part.
+file(REMOVE
+    "${ONNXIR_ROOT}/onnx/onnx.pb.h"
```
Comment (Yangqing) [inline CMakeLists.txt]: Naive question: do you want to remove headers?

Comment (linkerzhang) [reply CMakeLists.txt]: yes, this copy should be removed. This copy is generated by current "setup.py" call. if not removed, there will be two header/cc files in cmake projects. :)

File: CMakeLists.txt (hunk 19 above)
Comment (linkerzhang) [reply CMakeLists.txt]: thank you! Will add it.

File: onnx/defs/schema.cc [hunk 21]
```diff
@@ -115,22 +114,22 @@ void OpSchema::Verify(const NodeProto& node) const {
     }
   }
 
-  for (size_t out_idx = 0; out_idx < static_cast<size_t>(node.output_size());
```
Comment (larroy) [inline onnx/defs/schema.cc]: Why not leave size_t as the type?

Comment (linkerzhang) [reply onnx/defs/schema.cc]: Good point. You mean the return type of node.output_size(), right? will revisit the API later. This is fixing the warnings as I'm turning on taking warnings as errors. :)

**Comment by linkerzhang** [comment]
Thank you all for the comments! Most of the comments are resolved except two,
1. change setup.py, and 2) change the output_size() return type. Which will be covered/revisited in separate PRs. I'm now checking into it.

File: onnx/defs/schema.cc (hunk 21 above)
Comment (larroy) [reply onnx/defs/schema.cc]: yes, unintended narrowing casts can be a source of problems. I would leave them as size_t if possible.
